mlir-sycl/include/mlir/Dialect/SYCL/Analysis/ConstructorBaseAnalysis.h

```diff
@@ -0,0 +1,158 @@
+      return *static_cast<ConcreteAnalysis *>(this);
+    }
+
+    // Let the derived class perform more initialization of necessary.
+    static_cast<ConcreteAnalysis *>(this)->finalizeInitialization(
+        useRelaxedAliasing);
+
+    initialized = true;
+
+    return *static_cast<ConcreteAnalysis *>(this);
+  }
+
+  ConstructorBaseAnalysis(Operation *op, AnalysisManager &mgr)
+      : operation{op}, am{mgr}, solver{nullptr}, aliasAnalysis{nullptr} {}
+
+protected:
+  template <typename SYCLType>
+  std::optional<AnalysisResult> getInformationFromConstruction(Operation *op,
+                                                               Value operand) {
+    assert(initialized &&
+           "Analysis only available after successful initialization");
+    assert(isa<LLVM::LLVMPointerType>(operand.getType()) &&
+           "Expecting an LLVM pointer");
+    assert(aliasAnalysis != nullptr && "Alias analysis not initialized");
+
+    const polygeist::ReachingDefinition *reachingDef =
+        solver->lookupState<polygeist::ReachingDefinition>(op);
+    assert(reachingDef && "expected a reaching definition");
+
+    auto mods = reachingDef->getModifiers(operand, *solver);
+    if (!mods || mods->empty())
+      return std::nullopt;
+
+    if (!llvm::all_of(*mods, [&](const polygeist::Definition &def) {
+          return isConstructor<SYCLType>(def);
+        }))
+      return std::nullopt;
+
+    auto pMods = reachingDef->getPotentialModifiers(operand, *solver);
+    if (pMods) {
+      if (!llvm::all_of(*pMods, [&](const polygeist::Definition &def) {
+            return isConstructor<SYCLType>(def);
+          }))
+        return std::nullopt;
+    }
+
+    bool first = true;
+    AnalysisResult info;
+    for (const polygeist::Definition &def : *mods) {
+      if (first) {
+        info = getInformation<SYCLType>(def);
+        first = false;
+      } else {
+        info = info.join(getInformation<SYCLType>(def), *aliasAnalysis);
+        if (info.isTop())
+          // Early return: As soon as joining the different information has
+          // reached the top of the lattice, we can end processing.
+          return info;
+      }
+    }
+
+    if (pMods) {
+      for (const polygeist::Definition &def : *pMods) {
+        info = info.join(getInformation<SYCLType>(def), *aliasAnalysis);
+        if (info.isTop())
+          // Early return: As soon as joining the different information has
+          // reached the top of the lattice, we can end processing.
+          return info;
+      }
+    }
+
+    return info;
+  }
+
+  Operation *operation;
+
+  AnalysisManager &am;
+
+  std::unique_ptr<polygeist::DataFlowSolverWrapper> solver;
+
+  mlir::AliasAnalysis *aliasAnalysis;
+
+private:
+  template <typename SYCLType>
+  bool isConstructor(const polygeist::Definition &def) {
+    return static_cast<ConcreteAnalysis *>(this)
+        ->template isConstructorImpl<SYCLType>(def);
+  }
```

Comment:
From what I can see, `isConstructorImpl` bodies are the same for all `ConcreteAnalysis`. The only different is it's only defined for the types each analysis expects. Can we just drop `isConstructorImpl` and move the definition here?